### ISSUE TITLE
Processing English only archives

### ISSUE THREAD

**Issue by JaeHunRo** [commoncrawl/cc-pyspark#17]
First off, thanks for the great work on this data set.

I'd like to process archives that are English only. I saw that the columnar url index contained `content_languages` information (https://commoncrawl.s3.amazonaws.com/cc-index/table/cc-main/index.html). However, it seems that the index is for WARC files (though I understand that WET files are derived from WARC files). Thus, my initial plan was the following:

1. Use AWS Athena to select rows for a single month partition where `content_languages = "eng"` and save this as a separate table in my own s3 bucket.

```
CREATE TABLE "ccindex-2020-10-eng"
WITH (
  format = 'Parquet',
  external_location = 's3://my-bucket/ccindex-2020-10-eng',
)
AS SELECT url, warc_filename, warc_record_offset, warc_record_length
FROM "ccindex"."ccindex"
WHERE crawl = 'CC-MAIN-2020-10'
  AND subset = 'warc'
  AND content_languages = 'eng';
```

2. Use [cc_index_url_match.py](https://gist.github.com/JaeHunRo/a81fdb4b181d139cb6b731e57782c797) to process subset of WARC archives that are English only and use a regular expression to match for URLs in the document content that are camel case (e.g. `www.ThisIsATest.com`). The code is pretty much the same as cc-pyspark/cc_index_word_count.py with the exception of a few minor tweaks:

* Remove the `.persist()` on the input index (https://github.com/commoncrawl/cc-pyspark/blob/master/sparkcc.py#L329) since the index this time around is fairly large (>988 million rows)
* Change `output_schema` to just be a single key (the matched string found in document)
* Call `.distinct()` on RDD following `.mapPartitions` to remove duplicates
* Change regular expression

3. Run the Spark job (locally or on AWS EMR)
Local version (for debugging)

```
export SPARK_HOME=/some/path/to/spark-2.4.4-bin-hadoop2.7

$SPARK_HOME/bin/spark-submit \
  --packages com.amazonaws:aws-java-sdk:1.7.4,org.apache.hadoop:hadoop-aws:2.7.3 \
  ./cc_index_word_count.py \
  --query "SELECT url, warc_filename, warc_record_offset, warc_record_length FROM ccindex" \
  s3a://my-bucket/ccindex-2020-10-eng/ \
  output_table_name
```

I've also tried to set up the job to run on AWS EMR. The configuration I use on EMR is:
Hardware:
1 master = m5.xlarge (4 vCore, 16 GiB memory)
5 core = c5.4xlarge (16 vCore, 32 GiB memory)

Spark step:

```
spark-submit \
  --deploy-mode cluster \
  --packages com.amazonaws:aws-java-sdk-pom:1.10.34,org.apache.hadoop:hadoop-aws:2.7.2 \
  --py-files s3://my-bucket/sparkcc.py \
  s3://my-bucket/cc_index_url_match.py \
  --output_option path=s3://my-output-bucket/cc-main-2020-10-eng \
  --output_compression snappy --num_output_partitions 100 \
  --query "SELECT url, warc_filename, warc_record_offset, warc_record_length FROM ccindex"\
  s3a://my-bucket/ccindex-2020-10-eng/ \
  output_table_name
```

I use `--output_option path=s3://my-output-bucket/cc-main-2020-10-eng` to save the output table to my designated s3 bucket. However, this EMR job had numerous issues. The job has been running for >16 hours now, and it seems that the task metrics aren't being reported correctly since all tasks for the `saveAsTable` are marked as pending and no executors are shown to have any active tasks. It seems like it's stuck with ![all pending stages and no active stages](https://user-images.githubusercontent.com/8787226/82701385-1ad08500-9c3e-11ea-8c9b-1e7a197faca5.png) For more details, [eventLogs-application_1590106415063_0001-1.zip](https://github.com/commoncrawl/cc-pyspark/files/4669912/eventLogs-application_1590106415063_0001-1.zip).
As a sanity check, I used the exact same setup with a smaller version of the index with only 10 rows and it worked fine. 

To summarize:
* Is it just a bad idea to try to use an input index with this many rows?
* Is it possible to just use WET files somehow? 
(e.g. process WET files without worrying about `content_languages` and just produce `Row(match_url="www.ThisIsATest.com", target_uri="{record.rec_headers.get_header('WARC-Target-URI')}")` then JOIN this later with the index table on `target_uri = ccindex-2020-10-eng.url`)
* What is an efficient way to achieve the end result of processing English only content (w/o destroying my wallet)?

Thanks for your time.

**Comment by sebastian-nagel** [comment]
First, I would definitely not process one billion documents in a single job. It would require a larger cluster and if something goes wrong, all data extracted so far may get lost. Merging the output shouldn't be that difficult. There are many options how to split the input over jobs, eg. you could use the column "warc_segment" (there are 100 segments per crawl).

Second, if I understand the use case correctly, it's about extracting camel-case URLs. Just to be sure: is it intended to match URLs on the visible text of the page? There will be no match in `foo <a href="https://MyHost.com/">bar</a>` as the extracted text will be `foo bar`.

Of course, it would be possible to process the WET files directly. There is no language annotation there (at least, for now) but since a large portion of the content is English and if data from non-English pages isn't a problem, you'll be probably faster processing the WET files. If it's about links (not text), you could process the WAT files. Here, the detected language is store in the record after a response record.

**Comment by JaeHunRo** [comment]
Splitting into separate jobs is a good idea.
For the matching, it could be from visible text or in the raw HTML. 
Previously, I processed the WET files without considering content-language since it was faster but there were a non-trivial amount of non-English matches so I decided to revisit this. 
It sounds like processing WARC archives in separate jobs per `warc_segment` is the way to go since I'd like to process English-only content and visible text along with links.
